File: day3/part2/solution.rb
def find_max_subsequence(line, k)
  return "0" * k if line.length < k || k == 0

  result_chars = []
  current_start_index = 0

  (0...k).each do |i|
    remaining_to_find = k - i
    search_end_index = line.length - remaining_to_find

    best_digit = -1
    best_digit_index = -1
    
    # Find the best digit in the current search window
    (current_start_index..search_end_index).each do |j|
      digit = line[j].to_i
      if digit > best_digit
        best_digit = digit
        best_digit_index = j
      end
    end
    
    result_chars << best_digit.to_s
    current_start_index = best_digit_index + 1
  end

  result_chars.join
end

def solve
  input_path = File.join(File.dirname(__FILE__), 'input.txt')
  lines = File.readlines(input_path)

  total_joltage = 0
  k = 12 # Number of digits to select

  lines.each do |line|
    line = line.strip
    next if line.empty?

    max_line_joltage_str = find_max_subsequence(line, k)
    total_joltage += max_line_joltage_str.to_i
  end

  puts "The total output joltage is: #{total_joltage}"
end

solve

File: day3/part2/solution.js
const fs = require('fs');
const path = require('path');

function findMaxSubsequence(line, k) {
    if (line.length < k || k === 0) {
        return "0".repeat(k);
    }

    let resultChars = [];
    let currentStartIndex = 0;

    for (let i = 0; i < k; i++) {
        const remainingToFind = k - i;
        const searchEndIndex = line.length - remainingToFind;

        let bestDigit = -1;
        let bestDigitIndex = -1;

        for (let j = currentStartIndex; j <= searchEndIndex; j++) {
            const digit = parseInt(line[j], 10);
            if (digit > bestDigit) {
                bestDigit = digit;
                bestDigitIndex = j;
            }
        }
        
        resultChars.push(bestDigit.toString());
        currentStartIndex = bestDigitIndex + 1;
    }

    return resultChars.join('');
}


function solve() {
    const inputPath = path.join(__dirname, 'input.txt');
    const lines = fs.readFileSync(inputPath, 'utf-8').split('\n');

    let totalJoltage = 0n; // Use BigInt for total joltage
    const k = 12; // Number of digits to select

    for (const line of lines) {
        const trimmedLine = line.trim();
        if (!trimmedLine) {
            continue;
        }

        const maxLineJoltageStr = findMaxSubsequence(trimmedLine, k);
        totalJoltage += BigInt(maxLineJoltageStr);
    }

    console.log(`The total output joltage is: ${totalJoltage}`);
}

solve();
